### PR TITLE
Document mw extensions constraint

### PR DESCRIPTION
**Pull request by tzhelyazkova**

### CODE REVIEW

File: 02-Architecture Constraints.md [hunk 1]
```diff
@@ -10,4 +10,7 @@ Technical constraints that apply to the design and development of the system.
 
 Any Wikibase entity data displayed as part of article content has to be loadable by a [Lua](./12-Glossary.md/#Lua) function.
 
+## [MediaWiki Extensions](./12-Glossary.md/#mediawiki-extension )
 
+Any functionality that doesn't belong in [mediawiki core](https://www.mediawiki.org/wiki/Core) has to be registered as a mediawiki extension.
+**Mediawiki extensions cannot have extensions of their own.**
```
Comment (Silvan-WMDE) [inline 02-Architecture Constraints.md:16]: We briefly discussed this and I'm not sure there's a lot of value for us in this sentence. While technically everything is a Mediawiki extension, we can of course create an extension that provides additional functionality only to the Wikibase extension, so that it will have no effect if Wikibase is not there.

In fact, I think it would even help us to make the distinction between "Mediawiki extensions" and "Wikibase extensions", as Adam did in the list of components, even if technically there is no such thing.

**Review by Silvan-WMDE — COMMENTED** [review]
I vote for removing that one sentence, but wouldn't go as far as "requesting a change" here. :-)

File: 02-Architecture Constraints.md (hunk 1 above)
Comment (jakobw) [reply 02-Architecture Constraints.md:16]: +1

**Review by jakobw — COMMENTED** [review]
TL;DR lgtm but needs to be split with the client/repo split?

I agree that "has to be registered as a mediawiki extension" is a constraint for any client functionality we provide, because we can't (and don't want to) change the fact that they're MediaWiki applications. For Wikibase Repo, I don't think it's a constraint but rather a consequence of an early [architecture decision](https://docs.arc42.org/section-9/).

**Comment by addshore** [comment]
> because we can't (and don't want to) change the fact that they're MediaWiki applications

I feel like I challenge that (along the lines of something that I said in the call earlier).
There is no reason anything other than WIkibase Repo and WikibaseClient need to be MediaWiki extensions or to be registered and loaded in that way at all.
That is simply the way that we ended up doing things.

And with that in mind this is the sort of architectural change that we could make to dramatically simplify the relations and connections between these components (even if we then end up having ot maintain a bunch of stuff (interfaces etc) in both repo and client.

File: 02-Architecture Constraints.md (hunk 1 above)
Comment (addshore) [reply 02-Architecture Constraints.md:16]: I mean, they totally can? they just don't?

**Comment by jakobw** [comment]
> I feel like I challenge that (along the lines of something that I said in the call earlier).
> There is no reason anything other than WIkibase Repo and WikibaseClient need to be MediaWiki extensions or to be registered and loaded in that way at all.

Ah, but my point wasn't that we anything other than WikibaseRepo and WikibaseClient didn't need to be an extension. I tried to say:

1. WikibaseClient is an extension because it needs to directly integrate with MediaWiki applications and extend their functionality, which makes it a constraint.
2. WikibaseRepo is an extension because someone decided that it should be built on top of MediaWiki, which was an architectural decision, but not a constraint.

**Comment by addshore** [comment]
> > I feel like I challenge that (along the lines of something that I said in the call earlier).
> > There is no reason anything other than WIkibase Repo and WikibaseClient need to be MediaWiki extensions or to be registered and loaded in that way at all.
> 
> Ah, but my point wasn't that anything other than WikibaseRepo and WikibaseClient didn't technically need to be an extension. I tried to say:
> 
> 1. WikibaseClient is an extension because it needs to directly integrate with MediaWiki applications and extend their functionality, which makes it a constraint.
> 2. WikibaseRepo is an extension because someone decided that it should be built on top of MediaWiki, which was an architectural decision, but not a constraint.

+1 I think I totally agree with this then, and I think this is then what you expressed in `lgtm but needs to be split with the client/repo split?` ?

**Comment by jakobw** [comment]
Yup!

**Comment by Silvan-WMDE** [comment]
>     1. WikibaseClient is an extension because it needs to directly integrate with MediaWiki applications and extend their functionality, which makes it a constraint.
> 
>     2. WikibaseRepo is an extension because someone decided that it should be built on top of MediaWiki, which was an architectural decision, but not a constraint.

I fully agree.

**Comment by tzhelyazkova** [comment]
Closing due to some crazy rebase problems.
Will open a new PR.
